command/filemanager.go

```diff
@@ -53,3 +60,133 @@ func (fm *FileManager) CopyInlineFile(name string, fileContent pulumi.StringInpu
+}
+
+// CopyRelativeFile copies relative path to a remote path.
+// The relative path is defined in the same way as for `CopyRelativeFolder`.
+// This function returns the resource that can be used with `pulumi.DependsOn`.
+func (fm *FileManager) CopyRelativeFile(relativePath string, remotePath string, opts ...pulumi.ResourceOption) (pulumi.Resource, error) {
+	fullPath, _, err := getFullPath(relativePath, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	return fm.CopyFile(fullPath, remotePath, opts...)
+}
+
+// CopyFSFolder copies recursively a local folder to a remote folder.
+// You may consider using `CopyRelativeFolder` which has a simpler API.
+func (fm *FileManager) CopyFSFolder(
+	resourceName string,
+	folderFS fs.FS,
+	folderPath string,
+	remoteFolder string,
+	opts ...pulumi.ResourceOption) ([]pulumi.Resource, error) {
+	useSudo := true
+	folderCommand, err := fm.CreateDirectory(resourceName, pulumi.String(remoteFolder), useSudo, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create a temporary folder: %v for resource %v", err, resourceName)
+	}
+
+	files, folders, err := readFilesAndFolders(folderFS, folderPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read files and folders for %v", folderPath)
+	}
+
+	var folderResources []pulumi.Resource
+	for _, folder := range folders {
+		destFolder, err := getDestinationPath(folder, folderPath)
+		if err != nil {
+			return nil, err
+		}
+		remotePath := path.Join(remoteFolder, destFolder)
+		resources, err := fm.CreateDirectory("createFolder-"+remotePath, pulumi.String(remotePath), useSudo, pulumi.DependsOn([]pulumi.Resource{folderCommand}))
+		if err != nil {
+			return nil, err
+		}
+		folderResources = append(folderResources, resources)
+	}
+
+	fileResources := []pulumi.Resource{}
+	for _, file := range files {
+		destFile, err := getDestinationPath(file, folderPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		fileContent, err := fs.ReadFile(folderFS, file)
+		if err != nil {
+			return nil, err
+		}
+		fileCommand, err := fm.CopyInlineFile(
+			resourceName+"-"+file,
+			pulumi.String(fileContent),
+			path.Join(remoteFolder, destFile),
+			useSudo,
+			pulumi.DependsOn(folderResources))
+		if err != nil {
+			return nil, err
+		}
+		fileResources = append(fileResources, fileCommand)
+	}
+
+	return fileResources, nil
+}
+
+// When copying foo/bar to /tmp to result folder is /tmp/bar
+// This function remove the root prefix from the path (`foo` in this case)
+func getDestinationPath(folder string, rootFolder string) (string, error) {
+	baseFolder := filepath.Base(rootFolder)
+	rootWithoutBase := rootFolder[:len(rootFolder)-len(baseFolder)]
+
+	if !strings.HasPrefix(folder, rootWithoutBase) {
+		return "", fmt.Errorf("%v doesn't have the prefix %v", folder, rootWithoutBase)
+	}
+
+	return folder[len(rootWithoutBase):], nil
+}
+
+func getFullPath(relativeFolder string, skip int) (string, string, error) {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
What does `skip` do ?

Reply:
See https://pkg.go.dev/runtime#Caller.